File: python/modules/state/subband_erle.py
"""SubbandErleEstimator — per-bin ERLE with onset compensation (single-channel).

Mirrors docs/aec3_extracts/src/aec3/subband_erle_estimator.{cc,h}.

Single-channel port (num_capture_channels=1 always for us).

Rate notes:
  - ``kBlocksToHoldErle = 100`` AEC3 blocks (~400 ms) -> 40 hops.
  - ``kBlocksForOnsetDetection = 250`` AEC3 blocks (~1 s) -> 100 hops.
  - ``kPointsToAccumulate = 6`` is a count, not a time -> stays 6.
"""
import numpy as np

from ._constants import HOPS_PER_SECOND
from .. import aec3_scale as _aec3_scale

_AEC3_X2_BAND_ENERGY_THRESHOLD = 44015068.0  # AEC3 source value — scaled per hop in __init__
_BLOCKS_TO_HOLD_ERLE = int(0.4 * HOPS_PER_SECOND)            # AEC3 100 (~400 ms) -> 40 hops
_BLOCKS_FOR_ONSET_DETECTION = _BLOCKS_TO_HOLD_ERLE + int(0.6 * HOPS_PER_SECOND)  # AEC3 250 (~1 s) -> 100 hops
_POINTS_TO_ACCUMULATE = 6
_UNBOUNDED_ERLE_MAX = 100000.0

# AEC3 kFftLengthBy2Plus1 for the 16 kHz pipeline = 65 bins (FFT=128).
# Our pipeline has FFT=512 -> 257 bins. The first kFftLengthBy2 // 2 bins
# are the "LF half" that takes ``max_erle_l`` (AEC3 default 4.0); the rest
# take ``max_erle_h`` (default 1.5).


def _set_max_erle_bands(n_bins: int, max_erle_l: float, max_erle_h: float) -> np.ndarray:
    half = (n_bins - 1) // 2  # kFftLengthBy2 // 2 analogue
    arr = np.empty(n_bins, dtype=np.float32)
    arr[:half] = max_erle_l
    arr[half:] = max_erle_h
    return arr


class SubbandErleEstimator:
    def __init__(
        self,
        *,
        n_bins: int = 257,
        min_erle: float = 1.0,
        max_erle_l: float = 4.0,
        max_erle_h: float = 1.5,
        use_onset_detection: bool = True,
        use_min_erle_during_onsets: bool = True,
        hop_size: int = 160,
        e2y2_gate_enabled: bool = False,
        e2y2_gate_threshold: float = 0.5,
    ) -> None:
        self._n_bins = int(n_bins)
        self._min_erle = float(min_erle)
        self._max_erle = _set_max_erle_bands(self._n_bins, max_erle_l, max_erle_h)
        self._use_onset_detection = bool(use_onset_detection)
        self._use_min_erle_during_onsets = bool(use_min_erle_during_onsets)
        # AEC3 subband EMA constants are per-4ms-block; our update fires per hop.
        self._alpha_up = 0.05
        self._alpha_down = 0.1
        self._onset_release_decay = 0.97
        self._erle = np.full(self._n_bins, self._min_erle, dtype=np.float32)
        self._erle_onset_compensated = np.full(self._n_bins, self._min_erle, dtype=np.float32)
        self._erle_unbounded = np.full(self._n_bins, self._min_erle, dtype=np.float32)
        self._erle_during_onsets = np.full(self._n_bins, self._min_erle, dtype=np.float32)
        self._coming_onset = np.ones(self._n_bins, dtype=bool)
        self._hold_counters = np.zeros(self._n_bins, dtype=np.int32)
        # Accumulator state.
        self._y2_acc = np.zeros(self._n_bins, dtype=np.float32)
        self._e2_acc = np.zeros(self._n_bins, dtype=np.float32)
        self._low_render_energy = np.zeros(self._n_bins, dtype=bool)
        self._num_points = 0
        # C: E²/Y² per-bin ERLE gate (default OFF).
        # Skips ERLE update for bins where accumulated e2/y2 > threshold,
        # preventing nearend contamination in doubletalk from pulling ERLE down.
        self._e2y2_gate_enabled = bool(e2y2_gate_enabled)
        self._e2y2_gate_threshold = float(e2y2_gate_threshold)
        # Per-bin render energy floor — AEC3 kX2BandEnergyThreshold scaled to
        # our frame (2×hop_size actual samples, 50% OLA).
        self._x2_band_energy_threshold = _aec3_scale.per_bin_psd_threshold(
            _AEC3_X2_BAND_ENERGY_THRESHOLD, hop_size
        )
        self.reset()

    def reset(self) -> None:
        self._erle.fill(self._min_erle)
        self._erle_onset_compensated.fill(self._min_erle)
        self._erle_unbounded.fill(self._min_erle)
        self._erle_during_onsets.fill(self._min_erle)
        self._coming_onset.fill(True)
        self._hold_counters.fill(0)
        self._reset_accumulated_spectra()

    def update(
        self,
        *,
        x2: np.ndarray,
        y2: np.ndarray,
        e2: np.ndarray,
        converged_filter: bool,
        coh_gate_mask: np.ndarray = None,  # C': coherence gate (bool, len=n_bins)
    ) -> None:
        self._update_accumulated_spectra(x2, y2, e2, converged_filter)
        self._update_bands(converged_filter, coh_gate_mask=coh_gate_mask)
        if self._use_onset_detection:
            self._decrease_erle_per_band_for_low_render_signals()
        # Mirror first / last bin (AEC3 cc:100-109).
        self._erle[0] = self._erle[1]
        self._erle[-1] = self._erle[-2]
        self._erle_onset_compensated[0] = self._erle_onset_compensated[1]
        self._erle_onset_compensated[-1] = self._erle_onset_compensated[-2]
        self._erle_unbounded[0] = self._erle_unbounded[1]
        self._erle_unbounded[-1] = self._erle_unbounded[-2]

    def erle(self, onset_compensated: bool) -> np.ndarray:
        if onset_compensated and self._use_onset_detection:
            return self._erle_onset_compensated
        return self._erle

    def erle_unbounded(self) -> np.ndarray:
        return self._erle_unbounded

    def erle_during_onsets(self) -> np.ndarray:
        return self._erle_during_onsets

    # ----------------------------------------------------------- internals

    def _reset_accumulated_spectra(self) -> None:
        self._y2_acc.fill(0.0)
        self._e2_acc.fill(0.0)
        self._low_render_energy.fill(False)
        self._num_points = 0

    def _update_accumulated_spectra(
        self,
        x2: np.ndarray,
        y2: np.ndarray,
        e2: np.ndarray,
        converged_filter: bool,
    ) -> None:
        if not converged_filter:
            return
        if self._num_points == _POINTS_TO_ACCUMULATE:
            self._num_points = 0
            self._y2_acc.fill(0.0)
            self._e2_acc.fill(0.0)
            self._low_render_energy.fill(False)
        self._y2_acc += y2
        self._e2_acc += e2
        self._low_render_energy |= x2 < self._x2_band_energy_threshold
        self._num_points += 1

    def _update_bands(self, converged_filter: bool,
                      coh_gate_mask: np.ndarray = None) -> None:
        if not converged_filter:
            return
        if self._num_points != _POINTS_TO_ACCUMULATE:
            return
        new_erle = np.empty(self._n_bins, dtype=np.float32)
        is_erle_updated = np.zeros(self._n_bins, dtype=bool)
        # Per-bin compute only for k = 1 .. n_bins-2 (excludes endpoints).
        mid = slice(1, self._n_bins - 1)
        mask = self._e2_acc[mid] > 0.0
        new_erle[mid] = np.where(
            mask, self._y2_acc[mid] / np.maximum(self._e2_acc[mid], 1e-30), 0.0
        )
        is_erle_updated[mid] = mask
        # C: E²/Y² gate — freeze ERLE for bins where accumulated error energy
        # exceeds threshold × capture energy (nearend contaminating error in DT).
        # In FS (echo cancellation working): e2 << y2 → ratio low → not gated.
        # In DT: nearend inflates e2 toward y2 → ratio high → freeze ERLE update.
        if self._e2y2_gate_enabled:
            with np.errstate(divide='ignore', invalid='ignore'):
                e2y2 = self._e2_acc[mid] / np.maximum(self._y2_acc[mid], 1e-30)
            is_erle_updated[mid] &= (e2y2 <= self._e2y2_gate_threshold)
        # C': Coherence gate — freeze ERLE for bins where Γ²(Ŷ, Y) < threshold.
        # coh_gate_mask[k]=True means ERLE update allowed (coherence high enough).
        # In DT: nearend decorrelates Y from Ŷ → Γ² drops → False → freeze.
        # Unlike E²/Y² gate, this does NOT use E (error) → no circular dependency.
        if coh_gate_mask is not None:
            is_erle_updated[mid] &= coh_gate_mask[mid]
        # Onset detection: when bin sees "high enough" render AND was in
        # coming_onset state, transition out and hold for a window.
        if self._use_onset_detection:
            for k in range(1, self._n_bins - 1):
                if not is_erle_updated[k] or self._low_render_energy[k]:
                    continue
                if self._coming_onset[k]:
                    self._coming_onset[k] = False
                    if not self._use_min_erle_during_onsets:
                        alpha = 0.3 if new_erle[k] < self._erle_during_onsets[k] else 0.15
                        v = self._erle_during_onsets[k] + alpha * (
                            new_erle[k] - self._erle_during_onsets[k]
                        )
                        self._erle_during_onsets[k] = float(
                            np.clip(v, self._min_erle, self._max_erle[k])
                        )
                self._hold_counters[k] = _BLOCKS_FOR_ONSET_DETECTION
        # Per-bin smoothed ERLE update.
        for k in range(1, self._n_bins - 1):
            if not is_erle_updated[k]:
                continue
            low = bool(self._low_render_energy[k])
            self._erle[k] = self._smoothed_update(
                self._erle[k], new_erle[k], low, self._min_erle, float(self._max_erle[k])
            )
            if self._use_onset_detection:
                self._erle_onset_compensated[k] = self._smoothed_update(
                    self._erle_onset_compensated[k],
                    new_erle[k],
                    low,
                    self._min_erle,
                    float(self._max_erle[k]),
                )
            self._erle_unbounded[k] = self._smoothed_update(
                self._erle_unbounded[k],
                new_erle[k],
                low,
                self._min_erle,
                _UNBOUNDED_ERLE_MAX,
            )

    def _smoothed_update(
        self, prev: float, new_val: float, low_render: bool, min_v: float, max_v: float
    ) -> float:
        alpha = self._alpha_up
        if new_val < prev:
            alpha = 0.0 if low_render else self._alpha_down
        v = prev + alpha * (new_val - prev)
        return float(np.clip(v, min_v, max_v))

    def _decrease_erle_per_band_for_low_render_signals(self) -> None:
        for k in range(1, self._n_bins - 1):
            self._hold_counters[k] -= 1
            if self._hold_counters[k] <= (
                _BLOCKS_FOR_ONSET_DETECTION - _BLOCKS_TO_HOLD_ERLE
            ):
                if self._erle_onset_compensated[k] > self._erle_during_onsets[k]:
                    self._erle_onset_compensated[k] = max(
                        self._erle_during_onsets[k],
                        self._onset_release_decay * self._erle_onset_compensated[k],
                    )
                if self._hold_counters[k] <= 0:
                    self._coming_onset[k] = True
                    self._hold_counters[k] = 0

File: c_impl/src/subband_erle.c
/* subband_erle.c — C port of python/modules/state/subband_erle.py.
 *
 * Per-bin ERLE estimator with onset compensation (AEC3 SubbandErleEstimator,
 * single capture channel). See subband_erle.h for the parity contract.
 *
 * PRECISION: float32-by-design (f32 campaign — Python bit-exact parity
 * retired, drift accepted). All per-bin ERLE arithmetic is float32 end-to-end
 * (scalars combined with the f32 arrays round once, in float). The lone
 * threshold quantity _x2_band_energy_threshold is float32 too (narrowed once
 * from the double aec3_per_bin_psd_threshold() helper's result at init); the
 * low-render test `low |= x2(f32) < thr(f32)` is now a plain float32 compare.
 * Built with -ffp-contract=off so per-op float roundings are not fused.
 */
#include "subband_erle.h"

#include "aec3_scale.h"   /* aec3_per_bin_psd_threshold */

#include <string.h>

/* _set_max_erle_bands: arr[:half] = max_erle_l ; arr[half:] = max_erle_h
 * with half = (n_bins - 1) // 2  (kFftLengthBy2 // 2 analogue). */
static void set_max_erle_bands(float *arr, int n_bins,
                               float max_erle_l, float max_erle_h) {
    int half = (n_bins - 1) / 2;   /* Python floor div on non-negative */
    int k;
    for (k = 0; k < half; ++k) arr[k] = max_erle_l;
    for (k = half; k < n_bins; ++k) arr[k] = max_erle_h;
}

void subband_erle_init(SubbandErle *s, int n_bins,
                       float min_erle, float max_erle_l, float max_erle_h,
                       int use_onset_detection, int use_min_erle_during_onsets,
                       int hop_size,
                       float *max_erle_st, float *erle_st, float *erle_oc_st,
                       float *erle_unb_st, float *erle_during_st,
                       unsigned char *coming_onset_st, int32_t *hold_st,
                       float *y2_acc_st, float *e2_acc_st,
                       unsigned char *low_render_st) {
    s->n_bins = n_bins;
    s->min_erle = min_erle;
    s->use_onset_detection = use_onset_detection ? 1 : 0;
    s->use_min_erle_during_onsets = use_min_erle_during_onsets ? 1 : 0;
    s->alpha_up = 0.05f;
    s->alpha_down = 0.1f;
    s->onset_release_decay = 0.97f;
    /* _x2_band_energy_threshold = per_bin_psd_threshold(44015068.0, hop_size)
     * (ref_hop=160); float32 end-to-end. */
    s->x2_band_energy_threshold =
        aec3_per_bin_psd_threshold(SE_X2_BAND_ENERGY_THRESHOLD, hop_size, 160);

    s->max_erle = max_erle_st;
    s->erle = erle_st;
    s->erle_onset_compensated = erle_oc_st;
    s->erle_unbounded = erle_unb_st;
    s->erle_during_onsets = erle_during_st;
    s->coming_onset = coming_onset_st;
    s->hold_counters = hold_st;
    s->y2_acc = y2_acc_st;
    s->e2_acc = e2_acc_st;
    s->low_render_energy = low_render_st;

    set_max_erle_bands(s->max_erle, n_bins, max_erle_l, max_erle_h);
    subband_erle_reset(s);
}

void subband_erle_reset(SubbandErle *s) {
    int k;
    for (k = 0; k < s->n_bins; ++k) {
        s->erle[k] = s->min_erle;
        s->erle_onset_compensated[k] = s->min_erle;
        s->erle_unbounded[k] = s->min_erle;
        s->erle_during_onsets[k] = s->min_erle;
        s->coming_onset[k] = 1;
        s->hold_counters[k] = 0;
    }
    /* _reset_accumulated_spectra */
    for (k = 0; k < s->n_bins; ++k) {
        s->y2_acc[k] = 0.0f;
        s->e2_acc[k] = 0.0f;
        s->low_render_energy[k] = 0;
    }
    s->num_points = 0;
}

/* _smoothed_update(prev, new_val, low_render, min_v, max_v) -> float (f32).
 *
 * All-float32, single rounding per op:
 *   diff = new_val - prev          -> float32
 *   v    = prev + alpha * diff     -> float32
 *   v    = clip(v, min_v, max_v)   -> float32
 * (previously widened alpha*diff to float64 to mirror numpy's Python-float ×
 * f32-scalar promotion; that widen-then-narrow round trip is retired.) */
static float smoothed_update(const SubbandErle *s, float prev, float new_val,
                             int low_render, float min_v, float max_v) {
    float alpha = s->alpha_up;
    float diff = new_val - prev;                   /* f32 subtraction */
    float v;
    if (new_val < prev) {
        alpha = low_render ? 0.0f : s->alpha_down;
    }
    v = prev + alpha * diff;                       /* f32 */
    /* clip(v, min_v, max_v): clamp low then high. */
    if (v < min_v) v = min_v;
    if (v > max_v) v = max_v;
    return v;
}

/* _update_accumulated_spectra */
static void update_accumulated_spectra(SubbandErle *s,
                                       const float *x2, const float *y2,
                                       const float *e2, int converged_filter) {
    int k;
    if (!converged_filter) return;
    if (s->num_points == SE_POINTS_TO_ACCUMULATE) {
        s->num_points = 0;
        for (k = 0; k < s->n_bins; ++k) {
            s->y2_acc[k] = 0.0f;
            s->e2_acc[k] = 0.0f;
            s->low_render_energy[k] = 0;
        }
    }
    for (k = 0; k < s->n_bins; ++k) {
        s->y2_acc[k] = s->y2_acc[k] + y2[k];     /* f32 += f32 */
        s->e2_acc[k] = s->e2_acc[k] + e2[k];     /* f32 += f32 */
        /* low_render |= x2 < thr : both float32, compare directly. */
        if (x2[k] < s->x2_band_energy_threshold) {
            s->low_render_energy[k] = 1;
        }
    }
    s->num_points += 1;
}

/* _update_bands */
static void update_bands(SubbandErle *s, int converged_filter,
                         const unsigned char *coh_gate_mask) {
    int k;
    int n = s->n_bins;
    if (!converged_filter) return;
    if (s->num_points != SE_POINTS_TO_ACCUMULATE) return;

    /* Per-bin compute only for interior bins k = 1 .. n-2 (excludes endpoints).
     * We fold new_erle + is_erle_updated into a per-bin loop; for masked-out
     * bins new_erle is 0.0 (matches np.where else-branch) and is_erle_updated
     * is 0, so they never touch state. */
    for (k = 1; k <= n - 2; ++k) {
        float new_erle;
        int is_updated;
        float e2a = s->e2_acc[k];
        int low;
        /* mask = e2_acc > 0.0 ; new = where(mask, y2_acc/max(e2_acc,1e-30), 0) */
        if (e2a > 0.0f) {
            float denom = e2a > 1e-30f ? e2a : 1e-30f; /* np.maximum(e2_acc,1e-30) */
            new_erle = s->y2_acc[k] / denom;            /* f32 */
            is_updated = 1;
        } else {
            new_erle = 0.0f;
            is_updated = 0;
        }

        /* Coherence gate: is_updated &= coh_gate_mask[k]. */
        if (coh_gate_mask != NULL && is_updated) {
            if (!coh_gate_mask[k]) is_updated = 0;
        }

        low = s->low_render_energy[k] ? 1 : 0;

        /* Onset detection transition + hold. */
        if (s->use_onset_detection) {
            if (is_updated && !s->low_render_energy[k]) {
                if (s->coming_onset[k]) {
                    s->coming_onset[k] = 0;
                    if (!s->use_min_erle_during_onsets) {
                        /* v = during + alpha*(new_erle - during), all float32.
                         * alpha = 0.3f if new<during else 0.15f. */
                        float prev_d = s->erle_during_onsets[k];
                        float alpha = (new_erle < prev_d) ? 0.3f : 0.15f;
                        float diff = new_erle - prev_d;          /* f32 */
                        float v = prev_d + alpha * diff;         /* f32 */
                        float lo = s->min_erle;
                        float hi = s->max_erle[k];
                        if (v < lo) v = lo;            /* clip */
                        if (v > hi) v = hi;
                        s->erle_during_onsets[k] = v;
                    }
                }
                s->hold_counters[k] = SE_BLOCKS_FOR_ONSET_DETECTION;
            }
        }

        /* Per-bin smoothed ERLE update (only when is_erle_updated). */
        if (is_updated) {
            s->erle[k] = smoothed_update(s, s->erle[k], new_erle, low,
                                         s->min_erle, s->max_erle[k]);
            if (s->use_onset_detection) {
                s->erle_onset_compensated[k] =
                    smoothed_update(s, s->erle_onset_compensated[k], new_erle,
                                    low, s->min_erle, s->max_erle[k]);
            }
            s->erle_unbounded[k] =
                smoothed_update(s, s->erle_unbounded[k], new_erle, low,
                                s->min_erle, SE_UNBOUNDED_ERLE_MAX);
        }
    }
}

/* _decrease_erle_per_band_for_low_render_signals */
static void decrease_erle_for_low_render(SubbandErle *s) {
    int k;
    int n = s->n_bins;
    for (k = 1; k <= n - 2; ++k) {
        s->hold_counters[k] -= 1;
        if (s->hold_counters[k] <=
            (SE_BLOCKS_FOR_ONSET_DETECTION - SE_BLOCKS_TO_HOLD_ERLE)) {
            if (s->erle_onset_compensated[k] > s->erle_during_onsets[k]) {
                /* max(during, 0.97f * onset_compensated), all float32. */
                float decayed = s->onset_release_decay *
                                s->erle_onset_compensated[k];
                float during = s->erle_during_onsets[k];
                s->erle_onset_compensated[k] =
                    decayed > during ? decayed : during;
            }
            if (s->hold_counters[k] <= 0) {
                s->coming_onset[k] = 1;
                s->hold_counters[k] = 0;
            }
        }
    }
}

void subband_erle_update(SubbandErle *s,
                         const float *x2, const float *y2, const float *e2,
                         int converged_filter,
                         const unsigned char *coh_gate_mask) {
    int n = s->n_bins;
    update_accumulated_spectra(s, x2, y2, e2, converged_filter);
    update_bands(s, converged_filter, coh_gate_mask);
    if (s->use_onset_detection) {
        decrease_erle_for_low_render(s);
    }
    /* Mirror first / last bin (AEC3 cc:100-109). */
    s->erle[0] = s->erle[1];
    s->erle[n - 1] = s->erle[n - 2];
    s->erle_onset_compensated[0] = s->erle_onset_compensated[1];
    s->erle_onset_compensated[n - 1] = s->erle_onset_compensated[n - 2];
    s->erle_unbounded[0] = s->erle_unbounded[1];
    s->erle_unbounded[n - 1] = s->erle_unbounded[n - 2];
}

const float *subband_erle_erle(const SubbandErle *s, int onset_compensated) {
    if (onset_compensated && s->use_onset_detection) {
        return s->erle_onset_compensated;
    }
    return s->erle;
}

const float *subband_erle_erle_unbounded(const SubbandErle *s) {
    return s->erle_unbounded;
}

const float *subband_erle_erle_during_onsets(const SubbandErle *s) {
    return s->erle_during_onsets;
}
